KAFKA-6086: Provide for custom error handling when Kafka Streams fails to produce

farmdawgnation opened the pull request:
This PR creates and implements the `ProductionExceptionHandler` as described in [KIP-210](https://cwiki.apache.org/confluence/display/KAFKA/KIP-210+-+Provide+for+custom+error+handling++when+Kafka+Streams+fails+to+produce).

I've additionally provided a default implementation preserving the existing behavior. I fixed various compile errors in the tests that resulted from my changing of method signatures, and added tests to cover the new behavior.

farmdawgnation committed:
Implement ProductionExceptionHandler

This interface will be used to specify how to behave when there is an
exception while producing a result record to Kafka.

farmdawgnation committed:
Provide an AlwaysFailProductionExceptionHandler.

This is a default implementation of the production exception handler
that will always instruct Streams to fail when there is an error
producing.

This implementation is consistent with the behavior before KIP-210 was
implemented.

farmdawgnation committed:
Provide an AlwaysContinueProductionExceptionHandler

This production exception handler will always continue processing in
light of errors producing result records.

farmdawgnation committed:
Add config declaration for production exception handler

farmdawgnation committed:
Remove unused import

farmdawgnation committed:
Create the ProductionExceptionHandler, provide to RecordCollectorImpl

farmdawgnation commented:
From Jenkins:

```

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine java version from '9.0.1'.
```

Doesn't _look_ like I broke this, but rather may be an issue with the JDK9 job?

farmdawgnation commented:
This PR has been updated with additional test coverage and should be ready for review and merging if the [VOTE] thread that's currently active passes.

/cc @guozhangwang @tedyu @mjsax who participated in the discussion thread

tedyu commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -178,6 +180,11 @@
+     * {@code default.production.exception.handler}
+     */
+    private static final String DEFAULT_PRODUCTION_EXCEPTION_HANDLER_CLASS_CONFIG = "default.production.exception.handler";
+    private static final String DEFAULT_PRODUCTION_EXCEPTION_HANDLER_CLASS_DOC = "Exception handling class that implements thhe <code>org.apache.kafka.streams.errors.ProductionExceptionHandler</code> interface.";
```
Typo: thhe

farmdawgnation replied:
Good catch, will correct.

tedyu commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:
```diff
@@ -107,7 +114,9 @@ public void onCompletion(final RecordMetadata metadata, final Exception exceptio
                             offsets.put(tp, metadata.offset());
                         } else {
-                            if (sendException == null) {
+                            ProductionExceptionHandlerResponse exceptionAction = productionExceptionHandler.handle(serializedRecord, exception);
```
Looks like this call can be conditionally executed when sendException == null

farmdawgnation replied:
That would effectively mean that the exception only occurs on the first production exception that occurs.... Hmmm. Yes, I think that's the more sensible decision because otherwise a handler could end up in a situation where it returns `CONTINUE` but the app still falls over because of a prior situation where it returned `FAIL`.

Good catch, I'll fix this.

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
SUCCESS 
 8054 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
SUCCESS 
 8054 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
SUCCESS 
 8054 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
SUCCESS 
 8054 tests run, 5 skipped, 0 failed.
--none--

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -346,6 +353,11 @@
                     LogAndFailExceptionHandler.class.getName(),
                     Importance.MEDIUM,
                     DEFAULT_DESERIALIZATION_EXCEPTION_HANDLER_CLASS_DOC)
+            .define(DEFAULT_PRODUCTION_EXCEPTION_HANDLER_CLASS_CONFIG,
```
Nit: we order this by importance and parameter name. should go after `DEFAULT_KEY_SERDE_CLASS_CONFIG`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/errors/ProductionExceptionHandler.java:
```diff
@@ -0,0 +1,58 @@
+         */
+        public final int id;
+
+        ProductionExceptionHandlerResponse(final int id, final String name) {
```
Nit: single parameter per line formatting

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:
```diff
@@ -106,8 +113,10 @@ public void onCompletion(final RecordMetadata metadata, final Exception exceptio
-                        } else {
-                            if (sendException == null) {
+                        } else if (sendException == null) {
+                            ProductionExceptionHandlerResponse exceptionAction = productionExceptionHandler.handle(serializedRecord, exception);
```
Why do we call the handler only for the first send exception? Should we do the callback for each?

Nit: add `final` to `ProductionExceptionHandlerResponse exceptionAction`

Also, for `ProducerFencedException` we are doomed to fail for any future record, too. Should we call the handler for this case? Note, `ProducerFencedException` will result in a self-healing behavior and Streams will not die. This raises the question, if we should distinguish between fatal exceptions that will occur for all records (eg, authorization) and record specific errors (like record too large) and only call the handler it if makes sense to skip over a record?

farmdawgnation replied:
> Why do we call the handler only for the first send exception? Should we do the callback for each?

My thinking here was that once the decision had been made to `FAIL`, any further invocations of the handler would just be extra noise.

> Nit: add final to ProductionExceptionHandlerResponse exceptionAction

Sure

> Also, for ProducerFencedException we are doomed to fail for any future record, too. Should we call the handler for this case? Note, ProducerFencedException will result in a self-healing behavior and Streams will not die. This raises the question, if we should distinguish between fatal exceptions that will occur for all records (eg, authorization) and record specific errors (like record too large) and only call the handler it if makes sense to skip over a record?

I responded to this on the mailing list a bit, but you didn't mention `ProducerFencedException`.

I'm inclined to think that we should not invoke the handler for `ProducerFencedException` if it is always self-healing, but I did like the idea of giving the users hooks to cause their own side effects....

I need to think on this, but I'm really not thrilled with how this feels at the moment.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -649,7 +651,7 @@ RecordCollector recordCollector() {
     // visible for testing only
-    RecordCollector createRecordCollector(final LogContext logContext) {
-        return new RecordCollectorImpl(producer, id.toString(), logContext);
+    RecordCollector createRecordCollector(final LogContext logContext, final ProductionExceptionHandler productionExceptionHandler) {
```
nit: single parameter per line formatting

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -552,7 +553,7 @@ public void shouldFlushRecordCollectorOnFlushState() {
 
             @Override
-            RecordCollector createRecordCollector(final LogContext logContext) {
+            RecordCollector createRecordCollector(final LogContext logContext, final ProductionExceptionHandler productionExceptionHandler) {
```
Nit: formatting (same below)

mjsax reviewed (changes requested):
Thanks for the PR. Couple of comments.

Can you also rebase. Thx.

Commit:
Move AlwaysContinue into the test namespace

We're removing this from the public API to avoid it being used in
production, but it's still useful for testing

Commit:
Rename to DefaultProductionExceptionHandler

Commit:
Nuke some extra whitespace in StreamsConfig

Commit:
Correct compile errors in tests resulting from signature changes

Commit:
Integrate the productionExceptionHandler into the RecordCollectorImpl

Commit:
Resolve a boolean complexity error

Commit:
Add test coverage for different kinds of exception handlers

Commit:
Run non-fatal exceptions occuring immediately on send through handler

Commit:
Correctly order config settings

Commit:
Resolve formatting nitpicks from review

farmdawgnation commented:
Alright, this PR has been rebased and should now also be in line with the current state of KIP-210.

Commit:
Nuke a merge artifact that appeared in my PR

Commit:
Correct a typo in the production handler class doc

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:
```diff
@@ -146,17 +192,22 @@ public void onCompletion(final RecordMetadata metadata,
-                              topic,
-                              fatalException.getMessage()),
-                fatalException);
+        } catch (final Exception uncaughtException) {
```
`KafkaProducer.send()` can throw the following:
```
     * @throws IllegalStateException if a transactional.id has been configured and no transaction has been started
     * @throws InterruptException If the thread is interrupted while blocked
     * @throws SerializationException If the key or value are not valid objects given the configured serializers
     * @throws TimeoutException If the time taken for fetching metadata or allocating memory for the record has surpassed <code>max.block.ms</code>.
     * @throws KafkaException If a Kafka related error occurs that does not belong to the public API exceptions.
```
I think we need to stop the world for all of them and thus don't need to call the handler.

Maybe we should specify in the KIP, that the handler only applies to errors reported in the `Callback` but not to exceptions throws by `send()` directly ? \cc @guozhangwang @bbejeck @dguy WDYT?

farmdawgnation replied:
I'd be fine with that change — I thought there was a more diverse set of things that might fire back immediately when I did my first read-through.

Commit:
Do not invoke ProductionExceptionHandler for exceptions thrown directly

Per code review, we decided to not invoke the ProductionExceptionHandler
on Exceptions that are thrown directly from send, as all of those should
cause things to fall over immediately.

farmdawgnation commented:
@mjsax I've updated the PR accordingly.

farmdawgnation commented:
(And the KIP too)

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/errors/DefaultProductionExceptionHandler.java:
```diff
@@ -0,0 +1,37 @@
+import org.apache.kafka.clients.producer.ProducerRecord;
+
+/**
+ * Production exception handler that always instructs streams to fail when an exception
```
nit `Production exception handler` -> `{@code ProductionExceptionHandler}`

farmdawgnation replied:
That's reasonable.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:
```diff
@@ -41,18 +50,25 @@
     private final static String EXCEPTION_MESSAGE = "%sAbort sending since %s with a previous record (key %s value %s timestamp %d) to topic %s due to %s";
     private final static String PARAMETER_HINT = "\nYou can increase producer parameter `retries` and `retry.backoff.ms` to avoid this error.";
+    private final static String HANDLER_CONTINUED_MESSAGE = "Error sending records (key {} value {} timestamp {}) to topic {} due to {}; " +
+        "The exception handler chose to CONTINUE processing in spite of this error.";
```
super nit: instead of `exception handler chose to CONTINUE` maybe `exception handler is configured to CONTINUE` , but this could just be a personal preference.

farmdawgnation replied:
I prefer the more active `chose` because the `ProductionExceptionHandler` will take in the exception and make a decision on what to do for that particular exception. My expectation is that folks will white-list exceptions to ignore as they make sense for their implementation case.

bbejeck commented:
Thanks for the PR, left a couple of comments

mjsax reviewed (approved):
Thanks for the PR and KIP (and your patience @farmdawgnation -- great addition to Streams API!)

LGTM.

Call for review and merging @guozhangwang @dguy 

Note: KIP is not voted yet -- so don't merge before we can close the vote.

farmdawgnation committed:
Link to ProductionExceptionHandler in this javadoc

farmdawgnation commented:
@dguy @guozhangwang This KIP has been officially accepted, and this PR is ready for review and merging.

I checked out the test failure and it looks unrelated to what I did.

bbejeck commented:
retest this please

bbejeck commented:
Thanks for the PR, LGTM

farmdawgnation commented:
Updated the original description to be more accurate w.r.t. the current state of the PR

dguy commented:
Thaks @farmdawgnation, LGTM

dguy commented:
Merged to trunk

guozhangwang commented:
Thanks @farmdawgnation for the contribution!

mjsax commented:
@farmdawgnation Would you mind doing a follow up PR to update AK docs accordingly? We should mention this in the upgrade guide and also add content to the main documentation.

farmdawgnation commented:
Yeah, I can do that

farmdawgnation commented:
Hmmm I'm not actually sure where such a PR is supposed to go. The docs in this repo appear to be really old and the apache/kafka-site repo doesn't seem to have anything 1.1 related yet. Am I missing something?

mjsax commented:
PR should go against this repo/trunk; see `docs/` -- not sure why you think that those are old?

farmdawgnation commented:
I see what happened, I got confused. It looks like the upgrade guide hasn't gotten started for 1.1.0 yet, so I'm not sure what to add there. There aren't any changes required _to_ upgrade - the existing behavior is the default. But I can add a blurb about error handling to the main streams docs.
